File: Work_Api/LAMS.Logic/Services/Worker/WorkerService.cs
using AutoMapper;
using LAMS.DataAccess.Common.Models.Work;
using LAMS.DataAccess.Common.Repositories.Worker;
using LAMS.Logic.Common.Models.Work;
using LAMS.Logic.Common.Services.Worker;
using Spire.Doc;
using Spire.Pdf;
using Spire.Xls;
using Spire.Xls.Converter;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.Hosting;

namespace LAMS.Logic.Services.Worker
{
    public class WorkerService : IWorkerService
    {
        private IMapper _mapper;
        private IWorkerRepository _repo;

        public WorkerService(IWorkerRepository repo, IMapper mapper)
        {
            _repo = repo;
            _mapper = mapper;
        }

        public async Task<IEnumerable<OrderBLL>> GetNewOrders()
        {
            return await _repo.GetNewOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }

        public async Task<IEnumerable<OrderBLL>> GetProcessingOrders()
        {
            return await _repo.GetProcessingOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }

        public async Task<IEnumerable<OrderBLL>> GetWorkOrders()
        {
            return await _repo.GetWorkOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }

        public async Task<IEnumerable<OrderBLL>> GetWaitOrders()
        {
            return await _repo.GetWaitOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }
        public async Task<IEnumerable<OrderBLL>> GetApproveOrders()
        {
            return await _repo.GetApproveOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }
        public async Task<IEnumerable<OrderBLL>> GetFinishOrders()
        {
            return await _repo.GetFinishOrders()
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderBLL>>(t.Result));
        }
        public async Task<int> FinishOrder(string id)
        {
            return await _repo.FinishOrder(_mapper.Map<string>(id)).ContinueWith(t => t.Result);
        }
        public async Task<int> ProcessingOrder(string id)
        {
            return await _repo.ProcessingOrder(_mapper.Map<string>(id)).ContinueWith(t => t.Result);
        }
        public async Task<int> WaitOrder(string id)
        {
            return await _repo.WaitOrder(_mapper.Map<string>(id)).ContinueWith(t => t.Result);
        }
        public async Task<int> WorkOrder(string id)
        {
            return await _repo.WorkOrder(_mapper.Map<string>(id)).ContinueWith(t => t.Result);
        }

        public async Task<int> AddOrderMaterial(OrderMaterialBLL orderMaterial)
        {
            try
            {
                var id = await _repo.AddOrderMaterial(_mapper.Map<OrderMaterialDb>(orderMaterial)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public async Task<IEnumerable<OrderMaterialBLL>> GetOrderMaterials(string IdOrder)
        {
            return await _repo.GetOrderMaterials(IdOrder)
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderMaterialBLL>>(t.Result));
        }
        public async Task<OrderMaterialBLL> DelOrderMaterial(int Id)
        {
            return await _repo.DelOrderMaterial(Id)
                .ContinueWith(t => _mapper.Map<OrderMaterialBLL>(t.Result));
        }

        public async Task<int> AddOrderWorkType(OrderWorkTypeBLL orderWorkType)
        {
            try
            {
                var id = await _repo.AddOrderWorkType(_mapper.Map<OrderWorkTypeDb>(orderWorkType)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public async Task<IEnumerable<OrderWorkTypeBLL>> GetOrderWorkTypes(string IdOrder)
        {
            return await _repo.GetOrderWorkTypes(IdOrder)
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderWorkTypeBLL>>(t.Result));
        }
        public async Task<OrderWorkTypeBLL> DelOrderWorkType(int Id)
        {
            return await _repo.DelOrderWorkType(Id)
                .ContinueWith(t => _mapper.Map<OrderWorkTypeBLL>(t.Result));
        }

        public async Task<int> AddWorkerCost(OrderWorkerCostBLL orderWorkerCost)
        {
            try
            {
                var id = await _repo.AddWorkerCost(_mapper.Map<OrderWorkerCostDb>(orderWorkerCost)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public async Task<IEnumerable<OrderWorkerCostBLL>> GetWorkerCosts(string IdOrder)
        {
            return await _repo.GetWorkerCosts(IdOrder)
                .ContinueWith(t => _mapper.Map<IEnumerable<OrderWorkerCostBLL>>(t.Result));
        }
        public async Task<OrderWorkerCostBLL> DelWorkerCost(int Id)
        {
            return await _repo.DelWorkerCost(Id)
                .ContinueWith(t => _mapper.Map<OrderWorkerCostBLL>(t.Result));
        }
        public async Task<double> GetWorkerCostsSum(string IdOrder)
        {
            return await _repo.GetWorkerCostsSum(_mapper.Map<string>(IdOrder)).ContinueWith(t => t.Result);
        }
        public async Task<double> GetOrderMaterialsSum(string IdOrder)
        {
            return await _repo.GetOrderMaterialsSum(_mapper.Map<string>(IdOrder)).ContinueWith(t => t.Result);
        }
        public async Task<double> GetOrderWorkTypesSum(string IdOrder)
        {
            return await _repo.GetOrderWorkTypesSum(_mapper.Map<string>(IdOrder)).ContinueWith(t => t.Result);
        }
        public async Task<double> GetOneCost(int Id)
        {
            return await _repo.GetOneCost(_mapper.Map<int>(Id)).ContinueWith(t => t.Result);
        }
        public async Task<double> GetOneCostWorkType(int Id)
        {
            return await _repo.GetOneCostWorkType(_mapper.Map<int>(Id)).ContinueWith(t => t.Result);
        }
        public async Task<int> AddEvaluation(EvaluationBLL evaluation)
        {
            try
            {
                var id = await _repo.AddEvaluation(_mapper.Map<EvaluationDb>(evaluation)).ContinueWith(t => t.Result);

                return id;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }
        public async Task<EvaluationBLL> GetEvaluation(string Id)
        {
            return await _repo.GetEvaluation(Id)
                .ContinueWith(t => _mapper.Map<EvaluationBLL>(t.Result));
        }

        public async Task<int> AddTextAsync(DocumentBLL text)
        {
            return await _repo.AddTextAsync(_mapper.Map<DocumentDb>(text)).ContinueWith(t => t.Result);
        }
        public async Task<IEnumerable<DocumentBLL>> GetNamesOfTextsAsync(string IdOrder)
        {
            IEnumerable<DocumentBLL> textsInfo;

            textsInfo = await _repo.GetNamesOfTextsByIdAsync(IdOrder).ContinueWith(t => _mapper.Map<IEnumerable<DocumentBLL>>(t.Result));

            return textsInfo;
        }
        public DocumentDb GetTextById(int id)
        {
            return _repo.GetTextById(id);
        }
        #region IDisposable Support
        private bool disposedValue = false; // Для определения избыточных вызовов

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    // TODO: освободить управляемое состояние (управляемые объекты).
                }

                // TODO: освободить неуправляемые ресурсы (неуправляемые объекты) и переопределить ниже метод завершения.
                // TODO: задать большим полям значение NULL.

                disposedValue = true;
            }
        }

        // TODO: переопределить метод завершения, только если Dispose(bool disposing) выше включает код для освобождения неуправляемых ресурсов.
        // ~DocumentService() {
        //   // Не изменяйте этот код. Разместите код очистки выше, в методе Dispose(bool disposing).
        //   Dispose(false);
        // }

        // Этот код добавлен для правильной реализации шаблона высвобождаемого класса.
        void IDisposable.Dispose()
        {
            // Не изменяйте этот код. Разместите код очистки выше, в методе Dispose(bool disposing).
            Dispose(true);
            // TODO: раскомментировать следующую строку, если метод завершения переопределен выше.
            // GC.SuppressFinalize(this);
        }

        #endregion
    }
}

File: Work_UI/src/app/shared/worker.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams, HttpResponse } from '@angular/common/http';
import { environment } from 'src/environments/environment';
import { map } from 'rxjs/operators';
import { User, UserReg } from './user.model';
import { Observable, Subject } from 'rxjs';
import { SaveResult } from './save-result.model';
import { Order } from '../order/models/order.model';
import { OrderMaterial } from '../worker-ordermaterial/ordermaterial/ordermaterial.model';
import { OrderWorkType } from '../worker-orderworktype/models/orderworktype.model';
import { WorkerCost } from '../worker-orderworkercost/models/workercost.model';
import { Evaluation } from '../worker-evaluation/models/evaluation.model';
import { FileSystemFileEntry, UploadFile } from 'ngx-file-drop';
import { TextInfo } from '../order/models/text.model';

@Injectable({
  providedIn: 'root'
})
export class WorkerService {
  [x: string]: any;

  filesToSave: string[] = [];
  private URL_WORKER = `${environment.apiHostUrl}/api/worker`;

  constructor(
    private http: HttpClient
  ) { }

  saveTexts(id: string, files: UploadFile[]): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    let countRequetsts = 0;
    let countResponses = 0;
    const urlFile = this.URL_WORKER + '/uploadFiles';

    for (const entryFile of files) {
      const file = entryFile.fileEntry as FileSystemFileEntry;
      this.filesToSave.push(file.name);

      file.file((_file: File) => {
        const formData = new FormData();
        formData.append('IdOrder', id);
        formData.append('textsCount', files.length.toString());
        formData.append('textIndex', countRequetsts.toString());
        formData.append(_file.name, _file, _file.name);
        countRequetsts++;

        this.http.post(urlFile, formData).subscribe(
          res => {
            const index = this.filesToSave.indexOf(_file.name, 0);
            if (index > -1) {
              this.filesToSave.splice(index, 1);
            }
            countResponses++;
            if (countRequetsts === countResponses) {
              result.next(new SaveResult(true));
            }
          },
          err => {
            countResponses++;
            if (countRequetsts === countResponses) {
              result.next(new SaveResult(false));
            }
          });
      });
    }
    return result.asObservable();
  }
  getNamesOfTexts(IdOrder: string): Observable<TextInfo[]> {
    const url = `${environment.apiHostUrl}/api/worker/getnamesoftext?IdOrder=${IdOrder}`;
    return this.http.get<TextInfo[]>(url);
  }

  downloadTextById(id: number): Observable<HttpResponse<Blob>> {
    const url = `${environment.apiHostUrl}/api/worker/gettextbyid?Id=${id}`;
    let headers = new HttpHeaders();
    headers = headers.append('Accept', '*/*');

    return this.http.get(url, {
      headers: headers,
      observe: 'response',
      responseType: 'blob'
    });
  }
  getNewOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getneworders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  getProcessingOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getprocessingorders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  getWorkOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getworkorders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  getWaitOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getwaitorders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  getApproveOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getapproveorders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  getFinishOrders(): Observable<Order[]> {
    return this.http
      .get<Order[]>(this.URL_WORKER + '/getfinishorders')
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new Order(item.Id, item.UserId, item.Fio, item.Phone, item.Description, item.NeedEvaluation,
            item.Status));
        })
      );
  }

  FinishOrder(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/worker/finishorder?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }

  ProcessingOrder(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/worker/processingorder?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }

  WaitOrder(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/worker/waitorder?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }

  WorkOrder(id: string): Observable<SaveResult> {
    const url = `${environment.apiHostUrl}/api/worker/workorder?id=${id}`;
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    this.http.get(url).subscribe(
      res => {
        result.next();
      },
      err => {
        result.next();
      }
    );
    return result.asObservable();
  }
  AddMaterialOrder(newMaterial: OrderMaterial): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_WORKER + '/addordermaterial';

    this.http.post(urlDoc, newMaterial).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  getOrderMaterials(IdOrder): Observable<OrderMaterial[]> {
    const url = `${environment.apiHostUrl}/api/worker/getordermaterials?IdOrder=${IdOrder}`;
    return this.http
      .get<OrderMaterial[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new OrderMaterial(item.Id, item.IdMaterial, item.Material,
            item.IdOrder, item.Amount, item.Cost));
        })
      );
  }

  DeleteOrderMaterial(Id: string): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/worker/delordermaterial?Id=${Id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  AddWorkTypeOrder(newWorkType: OrderWorkType): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_WORKER + '/addorderworktype';

    this.http.post(urlDoc, newWorkType).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  getOrderWorkTypes(IdOrder): Observable<OrderWorkType[]> {
    const url = `${environment.apiHostUrl}/api/worker/getorderworktypes?IdOrder=${IdOrder}`;
    return this.http
      .get<OrderWorkType[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new OrderWorkType(item.Id, item.IdWorkType, item.WorkType,
            item.IdOrder,item.Amount, item.Cost));
        })
      );
  }

  DeleteOrderWorkType(Id: string): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/worker/delorderworktype?Id=${Id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  AddWorkerCost(newWorkerCost: WorkerCost): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_WORKER + '/addworkercost';

    this.http.post(urlDoc, newWorkerCost).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  getWorkerCostsSum(IdOrder): Observable<number> {
    const url = `${environment.apiHostUrl}/api/worker/getworkercostssum?IdOrder=${IdOrder}`;
    return this.http
      .get<number>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  getOneCost(IdMaterial): Observable<number> {
    const url = `${environment.apiHostUrl}/api/worker/getonecost?Id=${IdMaterial}`;
    return this.http
      .get<number>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  getOneCostWorkType(IdWorkType): Observable<number> {
    const url = `${environment.apiHostUrl}/api/worker/getonecostworktype?Id=${IdWorkType}`;
    return this.http
      .get<number>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  getOrderMaterialsSum(IdOrder): Observable<number> {
    const url = `${environment.apiHostUrl}/api/worker/getordermaterialssum?IdOrder=${IdOrder}`;
    return this.http
      .get<number>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  getOrderWorkTypesSum(IdOrder): Observable<number> {
    const url = `${environment.apiHostUrl}/api/worker/getorderworktypessum?IdOrder=${IdOrder}`;
    return this.http
      .get<number>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }
  getWorkerCosts(IdOrder): Observable<WorkerCost[]> {
    const url = `${environment.apiHostUrl}/api/worker/getworkercosts?IdOrder=${IdOrder}`;
    return this.http
      .get<WorkerCost[]>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result.map(item => new WorkerCost(item.Id, item.IdOrder, item.WorkersNumber, item.Cost));
        })
      );
  }

  DeleteWorkerCost(Id: string): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const url = `${environment.apiHostUrl}/api/worker/delworkercost?Id=${Id}`;

    this.http.get(url).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }
  AddEvaluation(evaluation: Evaluation): Observable<SaveResult> {
    const result: Subject<SaveResult> = new Subject<SaveResult>();
    const urlDoc = this.URL_WORKER + '/addevaluation';

    this.http.post(urlDoc, evaluation).subscribe(
      res => {
        if (res)
          result.next(new SaveResult(false));
        else
          result.next(new SaveResult(true));
      },
      error => {
        result.next(new SaveResult(true));
      });

    return result.asObservable();
  }

  getEvaluation(IdOrder): Observable<Evaluation> {
    const url = `${environment.apiHostUrl}/api/worker/getevaluation?Id=${IdOrder}`;
    return this.http
      .get<Evaluation>(url)
      .pipe(
        map(res => {
          const result: any = res;
          return result;
        })
      );
  }

  private getOptions(query: string, status: string) {
    return {
      params: new HttpParams()
        .set('UserId', query)
        .set('Status', status)
    };
  }
}
